src/github/client/github-client.ts

```diff
@@ -0,0 +1,93 @@
+import { Octokit } from "@octokit/rest";
+import axios, { AxiosInstance, AxiosRequestConfig } from "axios";
+import AppTokenCache from "./app-token-cache";
+import InstallationTokenCache from "./installation-token-cache";
+import AuthToken from "./auth-token";
+
+/**
+ * A GitHub client that supports authentication as a GitHub app.
+ *
+ * @see https://docs.github.com/en/developers/apps/building-github-apps/authenticating-with-github-apps
+ */
+export default class GitHubClient {
+
+	private readonly axios: AxiosInstance;
+	private readonly appTokenCache: AppTokenCache;
+	private readonly installationTokenCache: InstallationTokenCache;
+	private readonly githubInstallationId: number;
+
+	constructor(
+		appTokenCache: AppTokenCache,
+		installationTokenCache: InstallationTokenCache,
+		githubInstallationId: number,
+		baseURL = "https://api.github.com"
+	) {
+		this.axios = axios.create({
+			baseURL
+		});
+		this.appTokenCache = appTokenCache;
+		this.installationTokenCache = installationTokenCache;
+		this.githubInstallationId = githubInstallationId;
+	}
+
+	/**
+	 * Use this config in a request to authenticate with the app token.
+	 */
+	private asAppConfig(): Partial<AxiosRequestConfig> {
+		const appToken = this.appTokenCache.getAppToken();
+		return {
+			headers: {
+				Accept: "application/vnd.github.v3+json",
+				Authorization: `Bearer ${appToken.token}`
+			}
+		}
+	}
+
+	/**
+	 * Use this config in a request to authenticate with an installation token for the githubInstallationId.
+	 */
+	private async asInstallationConfig(): Promise<Partial<AxiosRequestConfig>> {
```

Comment:
Let's name it `getInstllationAuthenticationHeader` or something similar, which will reflect that the method generates authentication headers